endpoints/auction.go

```diff
@@ -482,3 +359,171 @@ func sortBidsAddKeywordsMobile(bids pbs.PBSBidSlice, pbs_req *pbs.PBSRequest, pr
+					Value:   bid.Adm,
+					IsVideo: true,
+				}
+			} else {
+				cobjs[i] = &pbc.CacheObject{
+					Value: &pbc.BidCache{
+						Adm:    bid.Adm,
+						NURL:   bid.NURL,
+						Width:  bid.Width,
+						Height: bid.Height,
+					},
+					IsVideo: false,
+				}
+			}
+		}
+		if err := pbc.Put(ctx, cobjs); err != nil {
+			return err
+		}
+		for i, bid := range resp.Bids {
+			bid.CacheID = cobjs[i].UUID
+			bid.CacheURL = a.cfg.GetCachedAssetURL(bid.CacheID)
+			bid.NURL = ""
+			bid.Adm = ""
+		}
+	} else if req.CacheMarkup == 2 {
+		return cacheVideoOnly(resp.Bids, ctx, a, labels)
+	}
+	return nil
+}
+
+func processBidResult(bidList pbs.PBSBidSlice, bidder *pbs.PBSBidder, aLabels *pbsmetrics.AdapterLabels, ctx *context.Context, req *pbs.PBSRequest, metrics pbsmetrics.MetricsEngine, ex adapters.Adapter) {
+	var err error
+	bidList, err = ex.Call(*ctx, req, bidder)
+	if err != nil {
+		var s struct{}
+		if err == context.DeadlineExceeded {
+			aLabels.AdapterErrors = map[pbsmetrics.AdapterError]struct{}{pbsmetrics.AdapterErrorTimeout: s}
+			bidder.Error = "Timed out"
+		} else if err != context.Canceled {
+			bidder.Error = err.Error()
+			switch err.(type) {
+			case *errortypes.BadInput:
+				aLabels.AdapterErrors = map[pbsmetrics.AdapterError]struct{}{pbsmetrics.AdapterErrorBadInput: s}
+			case *errortypes.BadServerResponse:
+				aLabels.AdapterErrors = map[pbsmetrics.AdapterError]struct{}{pbsmetrics.AdapterErrorBadServerResponse: s}
+			default:
+				glog.Warningf("Error from bidder %v. Ignoring all bids: %v", bidder.BidderCode, err)
+				aLabels.AdapterErrors = map[pbsmetrics.AdapterError]struct{}{pbsmetrics.AdapterErrorUnknown: s}
+			}
+		}
+	} else if bidList != nil {
+		bidList = checkForValidBidSize(bidList, bidder)
+		bidder.NumBids = len(bidList)
+		for _, bid := range bidList {
+			var cpm = float64(bid.Price * 1000)
+			metrics.RecordAdapterPrice(*aLabels, cpm)
+			switch bid.CreativeMediaType {
+			case "banner":
+				metrics.RecordAdapterBidReceived(*aLabels, openrtb_ext.BidTypeBanner, bid.Adm != "")
+			case "video":
+				metrics.RecordAdapterBidReceived(*aLabels, openrtb_ext.BidTypeVideo, bid.Adm != "")
+			}
+			bid.ResponseTime = bidder.ResponseTime
+		}
+	} else {
+		bidder.NoBid = true
+		aLabels.AdapterBids = pbsmetrics.AdapterBidNone
+	}
+	return
+}
+
+func (a *auction) recordMetrics(req *pbs.PBSRequest, labels pbsmetrics.Labels) {
+	if req == nil {
+		a.metricsEngine.RecordRequest(labels)
+		a.metricsEngine.RecordLegacyImps(labels, 0)
+	} else {
+		// handles the case that ParsePBSRequest returns an error, so req.Start is not defined
+		a.metricsEngine.RecordRequest(labels)
+		a.metricsEngine.RecordLegacyImps(labels, len(req.AdUnits))
+		a.metricsEngine.RecordRequestTime(labels, time.Since(req.Start))
+	}
+}
+
+func (a *auction) syncerProcess(req *pbs.PBSRequest, bidder *pbs.PBSBidder, blabels pbsmetrics.AdapterLabels, ex adapters.Adapter, ctx *context.Context) bool {
+	var skip bool = false
+	if req.App == nil {
```

Comment:
Isn't it more readable to have a single exit point at the end of the function whenever possible?

Reply:
I personally don't believe so. I think that was popular in C. The exit points should be clear, certainly. Its common to find "quick exits" at the start of the method to avoid indenting everything another level as each level of nesting generally makes the code harder to visually follow.

Reply:
Corrected